### ISSUE TITLE
Open Source booster notification rules

### ISSUE THREAD

**Issue by Ein-Tim** [corona-warn-app/cwa-wishlist#684]
## Feature description

The booster notification rules, which are currently effective, can't be seen publicly on GitHub (or on any other place known to me).

However, for transparency reasons, they should be made open source. This helps me and everyone else to better understand who's currently getting a notification and who's not. 

## Problem and motivation

As the situation reg. who is allowed to get a booster & what the STIKO recommends changes quite fast, it's currently very hard to keep up with in which scenarios a notification is send.

---
Internal Tracking ID: [EXPOSUREAPP-10984](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10984)

**Comment by svengabr** [comment]
Internal Tracking ID: EXPOSUREAPP-10984

**Comment by svengabr** [comment]
Thanks, @Ein-Tim, I did create a ticket and forwarded it to the team.

If the code cant be made public for some reason, I have asked to add it to the documentation.

**Comment by thomasaugsten** [comment]
You can see the rules at the moment here:
https://distribution.dcc-rules.de/bnrules/
https://distribution.dcc-rules.de/bnrules/31ab71cc2aa7e54e075b0b11d68b78fff33b12fbed185036d61b02649c3c63aa

**Comment by Ein-Tim** [comment]
For everybody interested in this, the booster notification rules are available here: https://distribution.dcc-rules.de/bnrules/

Just copy the hash of a rule & paste it after /bnrules/ and you will find the description & details of the rule. 

Please note that this is not the endpoint the app uses to download the rules, the URL the app is using is https://svc90.main.px.t-online.de/version/v1/booster-notification-rules.

Still, I would really like to see the booster notification rules published here on GitHub, so that it's very easy to find them.

Thanks to @vaubaehn for pointing me to the links I shared above!

**Comment by Ein-Tim** [comment]
Thanks to you @thomasaugsten & @svengabr!

**Comment by Ein-Tim** [comment]
Meanwhile, the BNR are very out of date (https://github.com/corona-warn-app/cwa-documentation/issues/754). 
I don't understand why they don't get updated or, if there is no intention to update them, removed.

If the rules where open-source, the community could suggest changes through PRs.

**Comment by vaubaehn** [comment]
Hi @dsarkar , could you please add this issue to EXPOSUREAPP-11340? Thank you!

@thomasaugsten do you see a chance that the pipeline for the booster notifications can be extended in such a way, that the JSONLogic of the rules gets its own repo? Then the community could propose changes to the rules there or even create PRs, so that changed rules can be merged and deployed after review.

**Comment by thomasaugsten** [comment]
In the current approval process it is not possible at the moment to integrate external sources.

**Comment by vaubaehn** [comment]
@thomasaugsten 
> In the current approval process it is not possible at the moment to integrate external sources.

Yes, but maybe in the future that could be made possible?
Same suggestion would be true for upcoming admission rules. I'm very optimistic that community can contribute and reduce delays.

**Comment by vaubaehn** [comment]
In a naive thinking, it would mean that the people who create/change the JSONLogic make their work files visible via a repo. If they need approval for their work, it could be sent from there. In this way, community could 'inject' their support.

**Comment by vaubaehn** [comment]
> In a naive thinking, it would mean that the people who create/change the JSONLogic make their work files visible via a repo. If they need approval for their work, it could be sent from there. In this way, community could 'inject' their support.

**Comment by MikeMcC399** [comment]
The process is currently intransparent. The rules are out of date and there is no feedback on the plans to implement newer rules:

- https://github.com/corona-warn-app/cwa-documentation/issues/754
- https://github.com/corona-warn-app/cwa-documentation/issues/806
- https://github.com/corona-warn-app/cwa-documentation/issues/814

**Comment by thomasaugsten** [comment]
We cannot give details about the process because we are not directly involved, please ask the publisher of the rules for details. 

To copy paste the content of the stiko text is only a minor task and will not speed up the process if is done by the community. 

I would be more helpful to have structured table of the 2G, 3G, 3G+, 2G+ and 2G+ + B definition for each federal state in Germany because my information this is not available at the moment and needed in many discussions

**Comment by MikeMcC399** [comment]
@thomasaugsten 

> We cannot give details about the process because we are not directly involved, please ask the publisher of the rules for details.

Can you tell us who the publisher of the rules is? How can we contact the publisher?

**Comment by vaubaehn** [comment]
Thanks for your response, @thomasaugsten . Just one last short question for now:
> We cannot give details about the process because we are not directly involved, please ask the publisher of the rules for details.

Are you actually talking about the publisher (RKI) or the company that hosts/deploys the booster notification rules to adress such a request?

> To copy paste the content of the stiko text is only a minor task and will not speed up the process if is done by the community.

I was thinking about taking care about the whole rule set (JSONlogic), not only altering the description in accordance with the STIKO recommendation. But you're right, for the booster notifications this is "überschaubar", but as we can see in practice, they're not maintained in the necessary speed. This is why I think, the community could help out here.

> I would be more helpful to have structured table of the 2G, 3G, 3G+, 2G+ and 2G+ + B definition for each federal state in Germany because my information this is not available at the moment and needed in many discussions

This is actually a problem. I don't know if there is an official institution at all to track this.
In worst case, maybe we can try to gather these information as a community effort - @Ein-Tim ?

Edit: Tim, I wasn't mentioning you to do the work, but to get your opinion. But if we try something like this, it could be helpful to ping your contacts on Twitter to engage in this work.

**Comment by Ein-Tim** [comment]
@vaubaehn 

> In worst case, maybe we can try to gather these information as a community effort

Phew, that's a hard thing especially as this changes nearly weekly. I would not be confident with even giving the rules for my own federal states at the moment. 😅

**Comment by vaubaehn** [comment]
> @vaubaehn
> 
> > In worst case, maybe we can try to gather these information as a community effort
> 
> Phew, that's a hard thing especially as this changes nearly weekly. I would not be confident with even giving the rules for my own federal states at the moment. 😅

@Ein-Tim 
Right, and me neither... But if we get some people from every federal state, that could be achieved, maybe. But I see the difficulties... But my gut feeling is, we won't get anything like this from RKI/BMG...

**Comment by vaubaehn** [comment]
Maybe we can put something like this on GH:
https://docs.google.com/spreadsheets/d/1W2Y0nUhRAwLr1LFEMIQogsIr3-08Is_FVPuAHKLiZY8/edit?usp=sharing

I will look whether in "Darf ich ~rein~ das?" information about admission rules is provided.

**Comment by thomasaugsten** [comment]
@vaubaehn Correct this are right contact for this RKI you can reach out impfnachweis@rki.de
I think if there a public repo available a lot people will help here to build up a long required list

**Comment by Ein-Tim** [comment]
@vaubaehn Please write me a DM via Slack, then we can talk about how we could start this.

**Comment by vaubaehn** [comment]
@thomasaugsten Thank you!
@Ein-Tim I will do!

**Comment by Bergreiter** [comment]
I like the idea 
I started a basic idea maybe this a good start
https://github.com/Bergreiter/ZugangsregelnDeutschland

**Comment by MikeMcC399** [comment]
@Ein-Tim 

Regarding the original request:
> The booster notification rules, which are currently effective, can't be seen publicly on GitHub (or on any other place known to me).

@dsarkar shared a script in https://github.com/corona-warn-app/cwa-documentation/issues/754#issuecomment-999644069 to extract the rules.

**Comment by vaubaehn** [comment]
Hi @Bergreiter and all,

thanks for your proposal regarding your repo [ZugangsregelnDeutschland](https://github.com/Bergreiter/ZugangsregelnDeutschland). Having a space on GitHub has advantages (public access, associated with the project) and disadvantages (concept of collaboration known to less users), and could be discussed further, though personally I tend to a GitHub-solution. In the end I think it would be good to have an official repo that is directly embedded into the CWA project, but accessable from the public. A table that we have started could result in the necessary admission rules finally. While trying a draft on a basic concept what information is needed to build up a table, for me it seems many information is nested:
- what certificates are necessary to apply for a specific admission rule (e.g., 2G +/B)
- which rules determine the validity of a specific certificate (e.g., what is 'boostered')
- which admission rules/certificate rules are valid for a specific federal state?

So we first need to discuss which information we need and how they could be ordered best to be able to fill out table cells in the end.

After I wrote above, I'll have a look into 'darf ich das', I had a look... And indeed some information are available, but they're not ordered in a comfortable way. I then saw, that the company maintaining the app is taking their data from Tourismus-wegweiser.de which is maintained by Project M GmbH Hamburg - and they're contracted by the Ministry of Economy and Energy. So maybe it's worth to officially ask there whether they already have structured data according admission rules and certificate rules? And connect to RKI to get a repo for admission rules as well as booster notification rules embedded into the CWA project?

Finally I suggest that we start a new issue for all further details about admission rules, as the issue here deals with booster notification rules.

How are your thoughts?

**Comment by vaubaehn** [comment]
@thomasaugsten and @mlenkeit 
The Kompetenzzentrum Tourismus des Bundes provides a public JSON-Api: https://tourismus-wegweiser.de/json
For sure, this data doesn't help. But it looks like they have tools to create this JSON... And if they have already structured data about admission rules, that could become a way to access these data also for CWA/CovPass(Check).
At least it's worth to engage in a joint effort/collaboration. Maybe you can ask RKI/BMG whether this is possible?

**Comment by dsarkar** [comment]
# German Acceptance Rules (Business Rules)

- **RR-DE-0001** Acceptance DE 1.0.0 1.0.0 CERTLOGIC 0.7.5 Recovery 2021-07-03T00:00:00Z 2030-06-01T00:00:00Z
  - **DE** Der positive NAA-Test (z.B. PCR) muss mindestens 28 Tage zurückliegen.
  - **EN** The positive NAA test result (e.g., PCR) must be older than 28 days.
  - [5a73c97d8c59495a5b48f3169e85d2cca0b4872548be21d9180f8c835c5004ba](https://distribution.dcc-rules.de/rules/DE/5a73c97d8c59495a5b48f3169e85d2cca0b4872548be21d9180f8c835c5004ba)
- **RR-DE-0002** Acceptance DE 1.0.1 1.0.0 CERTLOGIC 0.7.5 Recovery 2021-07-17T08:00:00Z 2030-06-01T00:00:00Z
  - **DE** Der positive NAA-Test (z.B. PCR) darf maximal 6 Monate zurückliegen.
  - **EN** The positive NAA test result (e.g., PCR) must be no older than 6 months.
  - [dc3afd7c87083a0f516e13a9ee26bb082e1deb119f94cb37c2de2ab63ba73bc2](https://distribution.dcc-rules.de/rules/DE/dc3afd7c87083a0f516e13a9ee26bb082e1deb119f94cb37c2de2ab63ba73bc2)
  - Issues: 
    - Should be no older than 90 days. 
    - Reference: https://www.rki.de/DE/Content/InfAZ/N/Neuartiges_Coronavirus/Genesenennachweis.html
    - Internal Tracking ID: [EXPOSUREAPP-11557](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-11557) update acceptance rule
    - Internal Tracking ID: [EXPOSUREAPP-11558](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-11558) create booster notification rule
- **TR-DE-0001** Acceptance DE 1.0.0 1.0.0 CERTLOGIC 0.7.5 Test 2021-07-03T00:00:00Z 2030-06-01T00:00:00Z
  - **DE** Es muss ein Antigen-Test (z.B. Schnelltest) oder NAA-Test (z.B. PCR) sein.
  - **EN** This must be an antigen test (e.g., rapid test) or NAA test (e.g., PCR).
  - [b3c4e2af3259bf5c793e0bed4ccf965ac54cb704a4949dfb55f4681d494a9a0e](https://distribution.dcc-rules.de/rules/DE/b3c4e2af3259bf5c793e0bed4ccf965ac54cb704a4949dfb55f4681d494a9a0e)
- **TR-DE-0002** Acceptance DE 1.0.0 1.0.0 CERTLOGIC 0.7.5 Test 2021-07-03T00:00:00Z 2030-06-01T00:00:00Z
  - **DE** Die Probenahme für einen Antigen-Test (z.B. Schnelltest) darf maximal 48 Stunden zurückliegen.
  - **EN** The sample for an antigen test (e.g., rapid test) must have been taken no longer than 48 hours ago.
  - [d43e24f487849ba261ca5c62c288482ccc454c75806acda550a442693bae8a1d](https://distribution.dcc-rules.de/rules/DE/d43e24f487849ba261ca5c62c288482ccc454c75806acda550a442693bae8a1d)
  - Issues:
    - 48 h is true when entering Germany, but within Germany 24 h for 3G and 2G+-events 
- **TR-DE-0003** Acceptance DE 1.0.0 1.0.0 CERTLOGIC 0.7.5 Test 2021-07-03T00:00:00Z 2030-06-01T00:00:00Z
  - **DE** Die Probenahme für einen NAA-Test (z.B. PCR) darf maximal 72 Stunden zurückliegen.
  - **EN** The sample for an NAA test (e.g., PCR) must have been taken no longer than 72 hours ago.
  - [130171fd2b90dadbd65c44650f60e715c59bfb2ca03913105a079c11d58b5803](https://distribution.dcc-rules.de/rules/DE/130171fd2b90dadbd65c44650f60e715c59bfb2ca03913105a079c11d58b5803)
  - Issues:
    - 48 h when entering Germany, and within Germany 48 h for 3G+ and 2G+-events 
- **TR-DE-0004** Acceptance DE 1.0.0 1.0.0 CERTLOGIC 0.7.5 Test 2021-07-03T00:00:00Z 2030-06-01T00:00:00Z
  - **DE** Das Ergebnis des Tests muss negativ sein.
  - **EN** The test result must be negative.
  - [34d53cfc0e330a05ddadb6262500e65a430a587b0dd29c027b5492f552e046d0](https://distribution.dcc-rules.de/rules/DE/34d53cfc0e330a05ddadb6262500e65a430a587b0dd29c027b5492f552e046d0)
- **TR-DE-0005** Acceptance DE 1.0.0 1.0.0 CERTLOGIC 0.7.5 Test 2021-07-03T00:00:00Z 2030-06-01T00:00:00Z
  - **DE** Der Antigen-Test (z.B. Schnelltest) muss in der Liste "Common list of rapid antigen tests" der EU enthalten sein.
  - **EN** The antigen test (e.g., rapid test) must be included in the EU’s "Common list of COVID-19 rapid antigen tests".
  - [7e177d5c072d0c1bdee4339e649e74ae5bf53198a4c44d615a203f98c57bbc1f](https://distribution.dcc-rules.de/rules/DE/7e177d5c072d0c1bdee4339e649e74ae5bf53198a4c44d615a203f98c57bbc1f)
- **VR-DE-0001** Acceptance DE 1.0.0 1.0.0 CERTLOGIC 0.7.5 Vaccination 2021-07-03T00:00:00Z 2030-06-01T00:00:00Z
  - **DE** Die Impfreihe muss vollständig sein (z.B. 1/1, 2/2).
  - **EN** The vaccination schedule must be complete (e.g., 1/1, 2/2).
  - [bc092f8000606c57a8fd80bc7a31ff720a4c4428510d88c774bed2f839c76b66](https://distribution.dcc-rules.de/rules/DE/bc092f8000606c57a8fd80bc7a31ff720a4c4428510d88c774bed2f839c76b66)
- **VR-DE-0002** Acceptance DE 1.0.0 1.0.0 CERTLOGIC 0.7.5 Vaccination 2021-07-03T00:00:00Z 2030-06-01T00:00:00Z
  - **DE** Nur die folgenden Impfstoffe werden akzeptiert: AstraZeneca, Biontech, Janssen, Moderna.
  - **EN** Only vaccines in the allowed valueset that have been approved by the EMA are allowed.
  - [dde2fb1ce39f854c8818ea56385d796ef84bd5573547aa12999f1239c38e04ef](https://distribution.dcc-rules.de/rules/DE/dde2fb1ce39f854c8818ea56385d796ef84bd5573547aa12999f1239c38e04ef)
- **VR-DE-0003** Acceptance DE 1.0.4 1.0.0 CERTLOGIC 0.7.5 Vaccination 2021-10-03T00:00:00Z 2030-06-01T00:00:00Z
  - **DE** Seit dem Abschluss der Grundimmunisierung müssen mindestens 14 Tage vergangen sein. Eine Auffrischungsimpfung oder eine Genesenenimpfung ist sofort gültig, falls sie eindeutig als solche erkennbar ist.
  - **EN** At least 14 days must have elapsed since completing the primary course of immunization. A booster shot or vaccination of someone who recovered from COVID-19 is valid immediately as long as it is clearly identified as such.
  - [f19e4792aeaffc8aba2ad59a851cbd3ceea6c0f44cafdb31cbeea6a6dcc7610e](https://distribution.dcc-rules.de/rules/DE/f19e4792aeaffc8aba2ad59a851cbd3ceea6c0f44cafdb31cbeea6a6dcc7610e)
- **VR-DE-0004** Acceptance DE 1.0.0 1.0.0 CERTLOGIC 0.7.5 Vaccination 2021-07-03T00:00:00Z 2030-06-01T00:00:00Z
  - **DE** Seit der Impfung darf maximal ein Jahr vergangen sein.
  - **EN** The vaccine must have been administered no longer than one year ago.
  - [1313e59f84f472694eece639cbe955b5c469ed7e0adb319a731d2d77f62f3e31](https://distribution.dcc-rules.de/rules/DE/1313e59f84f472694eece639cbe955b5c469ed7e0adb319a731d2d77f62f3e31)
  - Issues: 
    - From 1. Feb. 2020: The vaccine must have been administered no longer than 270 days ago.
    - Reference: https://ec.europa.eu/commission/presscorner/detail/de/ip_21_6837
   

# Booster Notification Rules

- **BNR-DE-0100** BoosterNotification DE 1.0.0 1.0.0 CERTLOGIC 1.0.0 General 2021-10-07T00:00:00Z 2030-06-01T00:00:00Z
  - **DE** Die Ständige Impfkommission (STIKO) empfiehlt allen Personen eine Auffrischimpfung, die mindestens 70 Jahre alt sind, bei denen keine Infektion mit dem Coronavirus SARS-CoV-2 nachgewiesen wurde und wenn ihre Grundimmunisierung über 6 Monate her ist.
Da Sie laut Ihrer gespeicherten Zertifikate bald dieser Personengruppe angehören und noch keine Auffrischimpfung erhalten haben, möchten wir Sie auf diese Empfehlung hinweisen. (Regel BNR-DE-0100)
Dieser Hinweis basiert ausschließlich auf den auf Ihrem Smartphone gespeicherten Zertifikaten. Die Verarbeitung der Daten erfolgte auf Ihrem Smartphone. Es wurden hierbei keine Daten an das RKI oder Dritte übermittelt.
  - **EN** The Standing Committee on Vaccination (STIKO) recommends that all persons who are 70 and over and who have not had a proven case of the coronavirus (SARS-CoV-2) have a booster shot if they obtained their base vaccination more than six months ago.
As your saved certificates indicate that you will soon belong to this group of persons and have not yet received a booster shot, we would like to inform you of this recommendation. (Rule BNR-DE-0100)
This notification is exclusively based on the certificates stored on your smartphone. The data was processed on your smartphone. No data was transferred to the RKI or third parties.
  - [31ab71cc2aa7e54e075b0b11d68b78fff33b12fbed185036d61b02649c3c63aa](https://distribution.dcc-rules.de/bnrules/31ab71cc2aa7e54e075b0b11d68b78fff33b12fbed185036d61b02649c3c63aa)
- **BNR-DE-0110** BoosterNotification DE 1.0.0 1.0.0 CERTLOGIC 1.0.0 General 2021-10-07T00:00:00Z 2030-06-01T00:00:00Z
  - **DE** Die Ständige Impfkommission (STIKO) empfiehlt allen Personen eine Auffrischimpfung, die mindestens 70 Jahre alt sind, bei denen keine Infektion mit dem Coronavirus SARS-CoV-2 nachgewiesen wurde und wenn ihre Grundimmunisierung über 6 Monate her ist.
Da Sie laut Ihrer gespeicherten Zertifikate dieser Personengruppe angehören und noch keine Auffrischimpfung erhalten haben, möchten wir Sie auf diese Empfehlung hinweisen. (Regel BNR-DE-0110)
Dieser Hinweis basiert ausschließlich auf den auf Ihrem Smartphone gespeicherten Zertifikaten. Die Verarbeitung der Daten erfolgte auf Ihrem Smartphone. Es wurden hierbei keine Daten an das RKI oder Dritte übermittelt.
  - **EN** The Standing Committee on Vaccination (STIKO) recommends that all persons who are 70 and over and who have not had a proven case of the coronavirus (SARS-CoV-2) have a booster shot if they obtained their base vaccination more than six months ago.
As your saved certificates indicate that you belong to this group of persons and have not yet received a booster shot, we would like to inform you of this recommendation. (Rule BNR-DE-0110)
This notification is exclusively based on the certificates stored on your smartphone. The data was processed on your smartphone. No data was transferred to the RKI or third parties.
  - [c0bd6b52cb26ff4fc8b1d2038537d8b3873cf03b23de50b7edcdb863c7fcd78a](https://distribution.dcc-rules.de/bnrules/c0bd6b52cb26ff4fc8b1d2038537d8b3873cf03b23de50b7edcdb863c7fcd78a)
- **BNR-DE-0200** BoosterNotification DE 1.0.0 1.0.0 CERTLOGIC 1.0.0 General 2021-10-07T00:00:00Z 2030-06-01T00:00:00Z
  - **DE** Die Ständige Impfkommission (STIKO) empfiehlt allen Personen eine weitere Impfstoffdosis zur Optimierung der Grundimmunisierung, die mit einer Dosis des Janssen-Impfstoffs (Johnson & Johnson) grundimmunisiert wurden, bei denen keine Infektion mit dem Coronavirus SARS-CoV-2 nachgewiesen wurde und wenn ihre Janssen-Impfung über 4 Wochen her ist.
Da Sie laut Ihrer gespeicherten Zertifikate bald dieser Personengruppe angehören und noch keine weitere Impfung erhalten haben, möchten wir Sie auf diese Empfehlung hinweisen. (Regel BNR-DE-0200)
Dieser Hinweis basiert ausschließlich auf den auf Ihrem Smartphone gespeicherten Zertifikaten. Die Verarbeitung der Daten erfolgte auf Ihrem Smartphone. Es wurden hierbei keine Daten an das RKI oder Dritte übermittelt.
  - **EN** The Standing Committee on Vaccination (STIKO) recommends that all persons who received a single dose of the Janssen vaccine (Johnson & Johnson) as their base vaccination and who have not had a proven case of the coronavirus (SARS-CoV-2) optimize their vaccination status with a further dose of a coronavirus vaccine if their Janssen vaccination was obtained more than four weeks ago.
As your saved certificates indicate that you will soon belong to this group of persons and have not yet received a further shot, we would like to inform you of this recommendation. (Rule BNR-DE-0200)
This notification is exclusively based on the certificates stored on your smartphone. The data was processed on your smartphone. No data was transferred to the RKI or third parties.
  - [a7285f30cb4d2a260ef678f6215225826b29315357d20f236bc54af23e2a1c8b](https://distribution.dcc-rules.de/bnrules/a7285f30cb4d2a260ef678f6215225826b29315357d20f236bc54af23e2a1c8b)
- **BNR-DE-0210** BoosterNotification DE 1.0.0 1.0.0 CERTLOGIC 1.0.0 General 2021-10-07T00:00:00Z 2030-06-01T00:00:00Z
  - **DE** Die Ständige Impfkommission (STIKO) empfiehlt allen Personen eine weitere Impfstoffdosis zur Optimierung der Grundimmunisierung, die mit einer Dosis des Janssen-Impfstoffs (Johnson & Johnson) grundimmunisiert wurden, bei denen keine Infektion mit dem Coronavirus SARS-CoV-2 nachgewiesen wurde und wenn ihre Janssen-Impfung über 4 Wochen her ist.
Da Sie laut Ihrer gespeicherten Zertifikate dieser Personengruppe angehören und noch keine weitere Impfung erhalten haben, möchten wir Sie auf diese Empfehlung hinweisen. (Regel BNR-DE-0210)
Dieser Hinweis basiert ausschließlich auf den auf Ihrem Smartphone gespeicherten Zertifikaten. Die Verarbeitung der Daten erfolgte auf Ihrem Smartphone. Es wurden hierbei keine Daten an das RKI oder Dritte übermittelt.
  - **EN** The Standing Committee on Vaccination (STIKO) recommends that all persons who received a single dose of the Janssen vaccine (Johnson & Johnson) as their base vaccination and who have not had a proven case of the coronavirus (SARS-CoV-2) optimize their vaccination status with a further dose of a coronavirus vaccine if their Janssen vaccination was obtained more than four weeks ago.
As your saved certificates indicate that you belong to this group of persons and have not yet received a further shot, we would like to inform you of this recommendation. (Rule BNR-DE-0210)
This notification is exclusively based on the certificates stored on your smartphone. The data was processed on your smartphone. No data was transferred to the RKI or third parties.
  - [ff4f8db6265f3d27bba3e0bd93893c6372c275da1d605ebb01f11e899de468ea](https://distribution.dcc-rules.de/bnrules/ff4f8db6265f3d27bba3e0bd93893c6372c275da1d605ebb01f11e899de468ea)
- **BNR-DE-0300** BoosterNotification DE 1.0.0 1.0.0 CERTLOGIC 1.0.0 General 2021-11-19T00:00:00Z 2022-04-30T00:00:00Z
  - **DE** Das Bundesministerium für Gesundheit (BMG) empfiehlt allen Personen, bei denen keine Infektion mit dem Coronavirus SARS-CoV-2 nachgewiesen wurde und deren Grundimmunisierung über 6 Monate her ist, einen Impftermin zu vereinbaren bzw. sich über den Nutzen einer Auffrischimpfung ärztlich aufklären zu lassen. Wissenschaftliche Studien belegen die höhere Schutzwirkung von Auffrischimpfungen vor einer COVID-19-Infektion.
Da Sie laut Ihrer gespeicherten Zertifikate dieser Personengruppe angehören und noch keine Auffrischimpfung erhalten haben, möchten wir Sie auf diese Empfehlung hinweisen. (Regel BNR-DE-0300)
Dieser Hinweis basiert ausschließlich auf den auf Ihrem Smartphone gespeicherten Zertifikaten. Die Verarbeitung der Daten erfolgte auf Ihrem Smartphone. Es wurden hierbei keine Daten an das RKI oder Dritte übermittelt.
  - **EN** The Federal Ministry of Health recommends that all persons who have not had a proven case of the coronavirus (SARS-CoV-2) and who obtained their base vaccination more than six months ago arrange a vaccination appointment or discuss the benefits of getting a booster shot with their doctor. Scientific studies confirm that booster shots provide a higher level of protection against becoming infected with COVID-19.
As your saved certificates indicate that you belong to this group of persons and have not yet received a booster shot, we would like to inform you of this recommendation. (Rule BNR-DE-0300)
This notification is exclusively based on the certificates stored on your smartphone. The data was processed on your smartphone. No data was transferred to the RKI or third parties.
  - [cd7defef43c2c0edb115090e669c2bb9028608e247c71b3d0f44728a3df4126f](https://distribution.dcc-rules.de/bnrules/cd7defef43c2c0edb115090e669c2bb9028608e247c71b3d0f44728a3df4126f)
- **BNR-DE-1000** BoosterNotification DE 1.0.0 1.0.0 CERTLOGIC 1.0.0 General 2021-11-05T00:00:00Z 2030-06-01T00:00:00Z
  - **DE** Aus Sicherheitsgründen müssen alle Zertifikate der von Ihnen besuchten Apotheke neu ausgestellt werden.
Ihr Zertifikat wird am 15.11.2021 gesperrt. Sie können sich unter Vorlage Ihres gelben Impfpasses und eines Lichtbildausweises in Ihrer Apotheke kostenfrei ein neues Impfzertifikat ausstellen lassen.
  - **EN** For security reasons, all certificates from the pharmacy you have visited must be reissued.
Your certificate will be blocked on November 15, 2021. You can arrange for a new vaccination certificate to be issued free of charge in your pharmacy by showing your yellow vaccination passport and a valid photo ID.
  - [b75771eb841a3e17205d590170ff30dd5998299cd67af72ce508f3d57062eda8](https://distribution.dcc-rules.de/bnrules/b75771eb841a3e17205d590170ff30dd5998299cd67af72ce508f3d57062eda8)

# Script - German Acceptance Rules (Business Rules)
```
for i in $( wget -qO- https://distribution.dcc-rules.de/rules/DE/   | jq -r '.[].hash' ) 
do 
  json=$(   wget -qO- https://distribution.dcc-rules.de/rules/DE/$i) 
  Identifier=$(     echo $json | jq -r '.Identifier')
  Type=$(           echo $json | jq -r '.Type')
  Country=$(        echo $json | jq -r '.Country')
  Version=$(        echo $json | jq -r '.Version')
  SchemaVersion=$(  echo $json | jq -r '.SchemaVersion')
  Engine=$(         echo $json | jq -r '.Engine')
  EngineVersion=$(  echo $json | jq -r '.EngineVersion')
  CertificateType=$(echo $json | jq -r '.CertificateType')
  ValidFrom=$(      echo $json | jq -r '.ValidFrom')
  ValidTo=$(        echo $json | jq -r '.ValidTo')
  desc_de=$(        echo $json | jq -r '.Description[1].desc')
  desc_en=$(        echo $json | jq -r '.Description[0].desc')

  echo "- **"$Identifier"**" $Type $Country $Version $SchemaVersion $Engine $EngineVersion $CertificateType $ValidFrom $ValidTo
  echo "  - **DE** $desc_de"
  echo "  - **EN** $desc_en"
  echo "  - [$i](https://distribution.dcc-rules.de/rules/DE/$i)"
done
```

# Script to get Booster Notification Rules

```
for i in $( wget -qO- https://distribution.dcc-rules.de/bnrules    | jq -r '.[].hash' ) 
do 
  json=$(   wget -qO- https://distribution.dcc-rules.de/bnrules/$i) 
  Identifier=$(     echo $json | jq -r '.Identifier')
  Type=$(           echo $json | jq -r '.Type')
  Country=$(        echo $json | jq -r '.Country')
  Version=$(        echo $json | jq -r '.Version')
  SchemaVersion=$(  echo $json | jq -r '.SchemaVersion')
  Engine=$(         echo $json | jq -r '.Engine')
  EngineVersion=$(  echo $json | jq -r '.EngineVersion')
  CertificateType=$(echo $json | jq -r '.CertificateType')
  ValidFrom=$(      echo $json | jq -r '.ValidFrom')
  ValidTo=$(        echo $json | jq -r '.ValidTo')
  desc_de=$(        echo $json | jq -r '.Description[1].desc' | sed /^$/d)
  desc_en=$(        echo $json | jq -r '.Description[0].desc' | sed /^$/d)

  echo "- **"$Identifier"**" $Type $Country $Version $SchemaVersion $Engine $EngineVersion $CertificateType $ValidFrom $ValidTo
  echo "  - **DE** $desc_de"
  echo "  - **EN** $desc_en"
  echo "  - [$i](https://distribution.dcc-rules.de/bnrules/$i)"
done

```

**Comment by vaubaehn** [comment]
Great work, @dsarkar .
I had a look, who probably have their hands on coding the business rules (via dgc repo - business rules test data) as they're similarly outdated (not yet adopted for rules taking effect Feb 1st) and as I assume the same persons code the booster notification rules.
The last change for a business rule was VR-DE-0003 in the test data repo... Maybe someone (else) can ping that person when it comes to questions regarding updating/correcting business/booster rules?

**Comment by vaubaehn** [comment]
Hi @thomasaugsten , Tim and me are going to prepare an eMail to RKI like you suggested in https://github.com/corona-warn-app/cwa-wishlist/issues/684#issuecomment-1020362571.
Would you be fine to be in cc, so that a response can directly reach you?

**Comment by thomasaugsten** [comment]
I'm not involved in the process no need to cc me here.

**Comment by vaubaehn** [comment]
Hi @thomasaugsten , one question regarding the booster notification rules:
Will the "old" booster-notification-rules-system be deprecated soon, and be merged with the "new CCL", or will BNRs and CCLs exist in parallel?

**Comment by thomasaugsten** [comment]
They will exist in parallel the client will download BNR and the CCL will apply the BNR on the existing certificates

**Comment by vaubaehn** [comment]
@thomasaugsten : Thank you - that information helps to refine the mail to RKI.

**Comment by MikeMcC399** [comment]
@vaubaehn 
Could you explain the term "CCL" for those (like me) not familiar with it? Is it a set of rules or is it a rule engine maybe?

**Comment by mlenkeit** [comment]
@MikeMcC399 
> Common Covid Logic (CCL) describes a layer of common logic - especially regarding the interpretation of Digital COVID Certificates (DCCs) - that can be shared across different operating systems and implementations.
> 
> The logic in CCL is described in JSON format. An engine processes the logic on the client against a defined input structure (e.g. the set of DCCs) and can expect a defined output structure. The client can update the JSON regularly from a server. This facilitates updates to CCL independent of client releases, e.g. in order to respond to legal or regulatory pandemic-related changes on short notice.

That's what the repo will say, will be available soon...

**Comment by MikeMcC399** [comment]
@mlenkeit 
Many thanks for educating us about the meaning and plans for CCL! Sounds excellent!

**Comment by Ein-Tim** [comment]
@MikeMcC399 

FYI: https://github.com/corona-warn-app/cwa-app-ccl

(cc @mlenkeit & @vaubaehn)

**Comment by vaubaehn** [comment]
@mlenkeit @thomasaugsten 
Don't have time, but couldn't wait to have a look - and it looks 🎉 and ❤️ !
Short question: Rule name/type "IR" -> is this for admission rules? ("Immunization Rule"?)

**Comment by Ein-Tim** [comment]
@thomasaugsten 

>  I would be more helpful to have structured table of the 2G, 3G, 3G+, 2G+ and 2G+ + B definition for each federal state in Germany because my information this is not available at the moment and needed in many discussions

Maybe this gives you a starting point: https://www.dehoga-corona.de/fileadmin/user_upload/UEbersicht_Bundeslaender_Ausnahme_von_Testpflicht__2G-Plus__insbesondere_fuer_Geboosterte_2022_01_26_12-00.pdf

**Comment by mlenkeit** [comment]
> Short question: Rule name/type "IR" -> is this for admission rules? ("Immunization Rule"?)

@vaubaehn _IR_ is the abbreviation for Invalidation Rule, a mechanism that allows to invalidate DCCs, e.g. "vaccination certificates issued by Issuer A between t0 and t1 are not valid".

**Comment by vaubaehn** [comment]
> > Short question: Rule name/type "IR" -> is this for admission rules? ("Immunization Rule"?)
> 
> @vaubaehn _IR_ is the abbreviation for Invalidation Rule, a mechanism that allows to invalidate DCCs, e.g. "vaccination certificates issued by Issuer A between t0 and t1 are not valid".

Good morning, @mlenkeit , thanks for your response. It's a bit confusing, as we had the rule type "Invalidation Rule" before, although it wasn't used, and there has not been any identifier for it.
I'm referring to
https://github.com/corona-warn-app/cwa-app-ccl/blob/2ae52b6d2b82131eda00984f8298e1062affddcb/resources/json-schema/dcc-validation-rule.json#L22-L36
where we see 'IR' as a rule identifier, and like 'Acceptance' (that never had any identifier) the 'Invalidation' is a rule type...
Is the "new" invalidation rule for upcoming feature of blocking fraudulent DCC hashes then?
And how are the 'admission rules' (rules to calculate an admission state relying on DccWalletInfo) entering the CCL? Are they coming as/together with rule type BNR?
Sorry, I hope I'm not stealing too much of your time.

**Comment by Ein-Tim** [comment]
This is quite a good start: https://github.com/corona-warn-app/dcc-rule-translation

Let's see how this evolves, if all notifications can be seen there I will close this issue.

**Comment by Ein-Tim** [comment]
This has been de-facto implemented with https://github.com/corona-warn-app/dcc-rule-translation. Therefore closing this issue.
